README.md
![](https://img.shields.io/badge/language-dart-orange.svg)
![](https://img.shields.io/badge/pub-v1.0.3-blue.svg)

`type_translator` can resolve any type that includes primitive and custom class transformation.

## Usage

```dart
import 'package:type_translator/type_translator.dart';
import 'package:type_token/type_token.dart';

main() {
  dynamic input;

  // -> int
  input = "1000";
  print(translateT<int>(input)); // print: 1000
  input = 2000;
  print(translateT<int>(input)); // print: 2000
  input = "apple";
//  print(translateT<int>(input)); // error

  // -> String
  input = 1000;
  print(translateT<String>(input)); // print: "1000"
  input = true;
  print(translateT<String>(input)); // print: "true"
  input = 50.8;
  print(translateT<String>(input)); // print: "50.8"

  // -> double
  input = 1000;
  print(translateT<double>(input)); // print: 1000.0
  input = "10.555";
  print(translateT<double>(input)); // print: 10.555

  // custom translator
  registerPrimitiveTranslatorT(customTranslator);
  input = A();
  print(translateT<int>(input)); // print: 90

  registerPrimitiveTranslatorT(customTranslator2);
  input = 1.55;
  print(translateT<int>(input));

  registerPrimitiveTranslatorT(customTranslator3);
  input = B(999);
  print(translateT<int>(input));
}

class A {}

class B {
  int data;

  B(this.data);
}

TypeTranslator<A, int> get customTranslator => TypeTranslator((input) => 90);

TypeTranslator get customTranslator2 => TypeTranslator.fromToken(TypeToken.ofDouble(), TypeToken.ofInt(), (input) => 10);

TypeTranslator get customTranslator3 => TypeTranslator.fromToken(TypeToken.of(B), TypeToken.ofInt(), (input) => input.data);
```

CHANGELOG.md
## 1.0.0

- Initial version, created by Stagehand

## 1.0.1

- new `translateInt`, `translateDouble`, `translateBool`, `translateString` 

## 1.0.2

- support `TypeToken` transform

## 1.0.3

- release default `translator`

lib/src/type_token.dart
class TypeToken {
  final String _typeName;

  TypeToken.parse(Object obj) : this.of(obj == null ? dynamic : obj.runtimeType);

  TypeToken.ofName(this._typeName);

  TypeToken.of(Type type) : this.ofName(type == null ? dynamic.toString() : type.toString());

  TypeToken.ofDynamic() : this.of(dynamic);

  TypeToken.ofInt() : this.of(int);

  TypeToken.ofString() : this.of(String);

  TypeToken.ofDouble() : this.of(double);

  TypeToken.ofBool() : this.of(bool);

  static TypeToken ofListByToken(TypeToken componentType) {
    return TypeToken.ofName('List<${componentType.typeName}>');
  }

  static TypeToken ofMapByToken(TypeToken keyType, TypeToken valueType) {
    return TypeToken.ofName('Map<${keyType.typeName}, ${valueType.typeName}>');
  }

  static TypeToken ofList<T>() {
    return ofListByToken(TypeToken.of(T));
  }

  static TypeToken ofMap<K, V>() {
    return ofMapByToken(TypeToken.of(K), TypeToken.of(V));
  }

  String get typeName => _typeName;

  bool get isPrimitive => ['int', 'double', 'bool', 'String'].contains(typeName);

  bool get isNotPrimitive => !isPrimitive;

  bool get isInt => typeName == 'int';

  bool get isDouble => typeName == 'double';

  bool get isBool => typeName == 'bool';

  bool get isString => typeName == 'String';

  bool get isList => RegExp('List(<.+>)?').hasMatch(typeName);

  bool get isMap => RegExp('Map(<.+, .+>)?').hasMatch(typeName);

  @override
  bool operator ==(Object other) =>
      identical(this, other) || other is TypeToken && runtimeType == other.runtimeType && _typeName == other._typeName;

  @override
  int get hashCode => _typeName.hashCode;
}

bool isPrimitive(Type type) => TypeToken.of(type).isPrimitive;

bool isInt(Type type) => TypeToken.of(type).isInt;

bool isDouble(Type type) => TypeToken.of(type).isDouble;

bool isBool(Type type) => TypeToken.of(type).isBool;

bool isString(Type type) => TypeToken.of(type).isString;

bool isList(Type type) => TypeToken.of(type).isList;

bool isMap(Type type) => TypeToken.of(type).isMap;


lib/src/type_translator_core.dart
import 'package:type_token/type_token.dart';

class TypeTranslators {
  // ======================================================================== primitive -> int

  static TypeTranslator<int, int> intToIntTranslator = TypeTranslator((input) => input);

  static TypeTranslator<String, int> stringToIntTranslator =
      TypeTranslator((input) => int.tryParse(input) ?? (throw _translationError(String, int)));

  static TypeTranslator<bool, int> boolToIntTranslator = TypeTranslator((input) => input ? 1 : 0);

// ======================================================================== primitive -> bool

  static TypeTranslator<bool, bool> boolToBoolTranslator = TypeTranslator((input) => input);

  static TypeTranslator<String, bool> stringToBoolTranslator = TypeTranslator((input) {
    var sBool = input.toLowerCase();
    return sBool == "true" ? true : sBool == "false" ? false : throw _translationError(String, bool);
  });

  static TypeTranslator<int, bool> intToBoolTranslator =
      TypeTranslator((input) => input == 1 ? true : input == 0 ? false : throw _translationError(int, bool));

// ======================================================================== primitive -> double

  static TypeTranslator<double, double> doubleToDoubleTranslator = TypeTranslator((input) => input);

  static TypeTranslator<String, double> stringToDoubleTranslator =
      TypeTranslator((input) => double.tryParse(input) ?? (throw _translationError(String, double)));

  static TypeTranslator<int, double> intToDoubleTranslator = TypeTranslator((input) => input.toDouble());

// ======================================================================== primitive -> String

  static TypeTranslator<int, String> intToStringTranslator = TypeTranslator((input) => input.toString());

  static TypeTranslator<double, String> doubleToStringTranslator = TypeTranslator((input) => input.toString());

  static TypeTranslator<bool, String> boolToStringTranslator = TypeTranslator((input) => input.toString());

  static TypeTranslator<String, String> stringToStringTranslator = TypeTranslator((input) => input);
}

String _translationErrorToken(TypeToken input, TypeToken expect) {
  return "Occur primitive translation error on ${input.fullTypeName} to ${expect.fullTypeName}";
}

String _translationError(Type input, Type expect) => _translationErrorToken(TypeToken.of(input), TypeToken.of(expect));

// ======================================================================== default translator factory

Map<TypeToken, _TypeTranslatorGroup> _translatorFactory = {
  TypeToken.ofInt(): _TypeTranslatorGroup<int>([
    TypeTranslators.intToIntTranslator,
    TypeTranslators.stringToIntTranslator,
    TypeTranslators.boolToIntTranslator,
  ]),
  TypeToken.ofDouble(): _TypeTranslatorGroup<double>([
    TypeTranslators.doubleToDoubleTranslator,
    TypeTranslators.intToDoubleTranslator,
    TypeTranslators.stringToDoubleTranslator,
  ]),
  TypeToken.ofBool(): _TypeTranslatorGroup<bool>([
    TypeTranslators.boolToBoolTranslator,
    TypeTranslators.intToBoolTranslator,
    TypeTranslators.stringToBoolTranslator,
  ]),
  TypeToken.ofString(): _TypeTranslatorGroup<String>([
    TypeTranslators.boolToStringTranslator,
    TypeTranslators.doubleToStringTranslator,
    TypeTranslators.intToStringTranslator,
    TypeTranslators.stringToStringTranslator,
  ]),
};

void registerPrimitiveTranslatorT<T, R>(TypeTranslator<T, R> translator) {
  TypeToken output = translator.outputType;
  if (!_translatorFactory.containsKey(output)) _translatorFactory[output] = _TypeTranslatorGroup<R>();
  _TypeTranslatorGroup group = _translatorFactory[output];
  group.add(translator);
}

void unregisterPrimitiveTranslatorByType(Type inputType, Type outputType) {
  TypeToken input = TypeToken.of(inputType);
  TypeToken output = TypeToken.of(outputType);
  if (_translatorFactory.containsKey(output)) {
    _TypeTranslatorGroup group = _translatorFactory[output];
    group.removeOnInput(input);
  }
}

void unregisterPrimitiveTranslator<T, R>() => unregisterPrimitiveTranslatorByType(T, R);

dynamic translateToken(dynamic input, TypeToken outputType) {
  if (input == null) return null;
  if (_translatorFactory.containsKey(outputType)) {
    _TypeTranslatorGroup group = _translatorFactory[outputType];
    return group.translate(input);
  }
  throw _translationErrorToken(TypeToken.of(input.runtimeType), outputType);
}

dynamic translate(dynamic input, Type outputType) => translateToken(input, TypeToken.of(outputType));

R translateT<R>(dynamic input) => translate(input, R);

int translateInt(dynamic input) => translateT<int>(input);

double translateDouble(dynamic input) => translateT<double>(input);

bool translateBool(dynamic input) => translateT<bool>(input);

String translateString(dynamic input) => translateT<String>(input);

// ======================================================================== define translator class

class TypeTranslator<T, R> {
  final R Function(T) _translationFunc;
  final TypeToken _inputType;
  final TypeToken _outputType;

  TypeTranslator(this._translationFunc)
      : _inputType = TypeToken.of(T),
        _outputType = TypeToken.of(R);

  TypeTranslator.fromToken(this._inputType, this._outputType, this._translationFunc);

  TypeToken get inputType => _inputType;

  TypeToken get outputType => _outputType;

  R translate(T input) => _translationFunc(input);
}

// ======================================================================== define translator group class

class _TypeTranslatorGroup<R> {
  final TypeToken outputType;
  Map<TypeToken, TypeTranslator> _translators = {};

  _TypeTranslatorGroup([List<TypeTranslator> initTranslators]) : outputType = TypeToken.of(R) {
    if (initTranslators != null) addAll(initTranslators);
  }

  void addT<T>(TypeTranslator<T, R> translator) {
    _translators[TypeToken.of(T)] = translator;
  }

  void add(TypeTranslator translator) {
    TypeToken inputType = translator.inputType;
    TypeToken outputType = translator.outputType;
    if (outputType != TypeToken.of(R)) throw "_TranslatorGroup<$R>.add<$outputType>() type must be same";
    _translators[inputType] = translator;
  }

  void addAll(List<TypeTranslator> translators) => translators.forEach((o) => add(o));

  void removeOnInput(TypeToken type) {
    if (_translators.containsKey(type)) {
      _translators.remove(type);
      print("unregister primitive translator $type -> $R");
    }
  }

  void removeOnInputType(Type type) => this.removeOnInput(TypeToken.of(type));

  R translate(dynamic input) {
    if (input == null) return null;
    TypeToken inputType = TypeToken.of(input.runtimeType);
    if (_translators.containsKey(inputType)) {
      TypeTranslator translator = _translators[inputType];
      return translator.translate(input);
    }
    throw _translationErrorToken(inputType, outputType);
  }
}


example/type_translator_example.dart
import 'package:type_translator/type_translator.dart';
import 'package:type_token/type_token.dart';

main() {
  dynamic input;

  // -> int
  input = "1000";
  print(translateT<int>(input)); // print: 1000
  input = 2000;
  print(translateT<int>(input)); // print: 2000
  input = "apple";
//  print(translateT<int>(input)); // error

  // -> String
  input = 1000;
  print(translateT<String>(input)); // print: "1000"
  input = true;
  print(translateT<String>(input)); // print: "true"
  input = 50.8;
  print(translateT<String>(input)); // print: "50.8"

  // -> double
  input = 1000;
  print(translateT<double>(input)); // print: 1000.0
  input = "10.555";
  print(translateT<double>(input)); // print: 10.555

  // custom translator
  registerPrimitiveTranslatorT(customTranslator);
  input = A();
  print(translateT<int>(input)); // print: 90

  registerPrimitiveTranslatorT(customTranslator2);
  input = 1.55;
  print(translateT<int>(input));

  registerPrimitiveTranslatorT(customTranslator3);
  input = B(999);
  print(translateT<int>(input));
}

class A {}

class B {
  int data;

  B(this.data);
}

TypeTranslator<A, int> get customTranslator => TypeTranslator((input) => 90);

TypeTranslator get customTranslator2 => TypeTranslator.fromToken(TypeToken.ofDouble(), TypeToken.ofInt(), (input) => 10);

TypeTranslator get customTranslator3 => TypeTranslator.fromToken(TypeToken.of(B), TypeToken.ofInt(), (input) => input.data);
